enhancements/cluster-logging/forward_to_cloudwatch.md

```diff
@@ -0,0 +1,249 @@
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+#### I want to group application logs by namespace
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+	   groupByNamespace: true
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+#### I want to define my own log groups (nice-to-have)
+
+There are two ways to define custom log groups:
+- Use a Kubernetes label with the group name.
+- Define the group name in the `ClusterLogForwarder.Output`
+
+1. Using a kubernetes label named `LogGroup`
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+	   groupByLabel: LogGroup
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+2. Defining a group in the output spec and use input selectors:
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+  inputs:
+  - name: FooIn
+    application:
+	  namespaces: [foo]
+  - name: BarIn
+    application:
+	  namespaces: [bar]
+  outputs:
+   - name: FooOut
+     type: cloudwatch
+     cloudwatch:
+	   group: foo
+   - name: BarOut
+     type: cloudwatch
+     cloudwatch:
+	   group: bar
+  pipelines:
+  - inputRefs: [FooIn]
+    outputRefs: [FooOut]
+  - inputRefs: [BarIn]
+    outputRefs: [BarOut]
+```
+
+#### I want to use EKS authentication with Cloudwatch
+
+TODO: is this a requirement? If so define an appropriate set of `secret` keys.
+
+### Implementation Details
+
+Using the [fluentd cloudwatch plugin][plugin].
+Most of the CRD fields have an obvious corresponding fluentd parameter.
+
+Set `auto_create_stream true` to create streams and groups on the fly.
+
+`log-stream-name` is set to the `containerID` of the log record.
+
+Log group names are set using `log_group_name` for a static `group` field.
+The `log_group_name_key` can select the log category, namespace or label as group name.
+
+### Open Questions
+
+Should the cluster logging operator need to delete log streams that are no longer in use? For example when containers are deleted?
+
+Is region a required parameter for connecting to cloudwatch? If not make that clear and evaluate if it's necessary.
+
+### Risks and Mitigations
+
+[Cloudwatch quota][quota] can be exceeded if insufficiently granular streams are configured.
+We configure a stream-per-container which is the finest granularity we have for logging.
```

Comment:
The throttling can also be "unfair" across nodes.  For large clusters, the risk that certain streams always fail is a real threat.

Having a central forwarder in the cluster would be one way to mitigate this.

Reply:
Do you have any pointers to cloudwatch's fairness strategy? Not clear to me how a central forwarder would mitigate this; a central forwarder would need to replicated, and to split traffic into small log streams given the per-stream quotas; so it would look pretty the same to cloudwatch - a small set of replicas generating a large set of log streams, vs. a small set of nodes generating a large set of log streams.

Reply:
Under the assumption that no node in the cluster shares the same log stream with another node (stream ID is either container UUID or is the unique host name), then the only 800/1,500 transactions a second quota can be hit (too many "small" PutLogEvent requests combined).

For a 50 node cluster, that would be 16/30 PutLogEvent reqs / sec at 1 MB per req.

So a really chatty node can chew up that up (that is only 16/30 MB/s).

Reply:
Understood, but there's no way we can work around that. If excessively chatty log streams use up a  cluster-wide quota, then others will starve. Doesn't matter if the streams originate from nodes or from replicas of a central forwarder.

Reply:
Rate limiting is still in our plans, when we implement that we'll be able to enforce some policy. Until then we don't control the volume of traffic.

Reply:
But by having central nodes, you can ensure the throttling is spread out properly, and no one node gets it worse than another, where as you are beholding to how CloudWatch throttles without it.